services/slam/builtin/cartographer_int_test.go

```diff
@@ -0,0 +1,150 @@
+	if err := os.Mkdir(name+"/config/lua_files", os.ModePerm); err != nil {
+		return err
+	}
+	for _, luaFile := range []string{"locating_in_map.lua", "mapping_new_map.lua", "updating_a_map.lua"} {
+		source, err := os.Open(artifact.MustPath("slam/" + luaFile))
+		if err != nil {
+			return err
+		}
+		defer source.Close()
+		destination, err := os.Create(name + "/config/lua_files/" + luaFile)
+		if err != nil {
+			return err
+		}
+		defer destination.Close()
+		_, err = io.Copy(destination, source)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Checks the cartographer position and map.
+func testCartographerPositionAndMap(t *testing.T, svc slam.Service) {
+	t.Helper()
+
+	actualMIME, _, pointcloud, err := svc.GetMap(context.Background(), "test", "pointcloud/pcd", nil, false)
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, actualMIME, test.ShouldResemble, "pointcloud/pcd")
+	test.That(t, pointcloud.Size(), test.ShouldEqual, 200)
+
+	// TODO DATA-701 test GetPosition
+}
+
+func TestCartographerIntegration(t *testing.T) {
+	_, err := exec.LookPath("carto_grpc_server")
+	if err != nil {
+		t.Log("Skipping test because carto_grpc_server binary was not found")
+		t.Skip()
+	}
+
+	name, err := createTempFolderArchitecture()
+	test.That(t, err, test.ShouldBeNil)
+	createLuaFiles(name)
+
+	t.Log("Testing online mode")
+
+	attrCfg := &builtin.AttrConfig{
+		Algorithm: "cartographer",
+		Sensors:   []string{"cartographer_int_lidar"},
+		ConfigParams: map[string]string{
+			"mode": "2d",
+			"v":    "1",
+		},
+		DataDirectory: name,
+	}
+
+	// Release point cloud for service validation
+	cartographerIntLidarReleasePointCloudChan <- 1
+	// Create slam service using a real cartographer binary
+	svc, err := createSLAMService(t, attrCfg, golog.NewTestLogger(t), true, true)
+	test.That(t, err, test.ShouldBeNil)
+
+	// Release point cloud, since cartographer looks for the second most recent point cloud
+	cartographerIntLidarReleasePointCloudChan <- 1
+	// Wait for cartographer to finish processing data
+	logReader := svc.(internal.Service).GetSLAMProcessBufferedLogReader()
+	for i := 0; i < numCartographerPointClouds-2; i++ {
+		t.Logf("Find log line for point cloud %v", i)
+		cartographerIntLidarReleasePointCloudChan <- 1
+		for {
+			line, err := logReader.ReadString('\n')
+			test.That(t, err, test.ShouldBeNil)
+			if strings.Contains(line, "Passed sensor data to SLAM") {
+				break
+			}
+		}
+	}
+
+	testCartographerPositionAndMap(t, svc)
+
+	// Close out slam service
+	test.That(t, utils.TryClose(context.Background(), svc), test.ShouldBeNil)
+	// Don't clear out the directory, since we will re-use the data for the next run
+	closeOutSLAMService(t, "")
+
+	// Delete the last point cloud, so that offline mode runs on the same data
+	files, err := ioutil.ReadDir(name + "/data/")
+	test.That(t, err, test.ShouldBeNil)
+	lastFileName := files[len(files)-1].Name()
+	test.That(t, os.Remove(name+"/data/"+lastFileName), test.ShouldBeNil)
```

Comment:
[q] When you say "delete the last point cloud", what point cloud are you referring to? To the last .pcd file in the `data/data` directory? Or the latest generated map, which would be a pbstream? Latter would make sense to me but that's not what the code seems to be doing, so I'm curious to understand this part better.

Reply:
I meant the last .pcd file in the `data/data` directory. I'll change the comment to make that more clear. The reason to delete the last .pcd file is so that online and offline mode run on the same set of data. Online mode will not read the last .pcd file, since it always processes the second-most-recent .pcd file, in case the most-recent .pcd file is currently being written.

Reply:
ah that's helpful, thank you!